### DISCUSSION TITLE
How to use transactions while offline

### DISCUSSION THREAD

**Discussion by BillyCottrell** [getsentry/sentry-java#2303] [Q&A] [answered]
Hi,

I am currently making a room booking system for a specific type of device in android and this app runs constantly on this device in kiosk mode. Now everytime it reboots it starts the app back up but at the start there is no internet (ethernet connection) hence I created a network check before continuing the app. Problem is that on first setup I usually request app permissions first before I check the network so the user will have to wait less.

Now I am wondering if it is possible to create a transaction to ensure that the app permission check worked and it didn't crash the app (I am requesting this as I had issues with it before as the app was not responding anymore).
Also as a side request is there a way to track amount of times network check occured or how long it took till network was available?

Any ideas?

Thanks in advance!

**Comment by philipphofmann** [comment, marked as answer]
> Now I am wondering if it is possible to create a transaction to ensure that the app permission check worked and it didn't crash the app

I'm unsure if transactions are the right way to do this. If the app crashes, our SDK will create a crash event and send it to Sentry. 

> Also as a side request is there a way to track amount of times network check occured or how long it took till network was available?

What you describe looks like a use case for custom metrics for performance to me. You could attach both metrics you described as measurements to a transaction, see https://docs.sentry.io/platforms/android/performance/instrumentation/performance-metrics/.

**Reply by BillyCottrell** [reply]
> I'm unsure if transactions are the right way to do this. If the app crashes, our SDK will create a crash event and send it to Sentry.

Well, the reason I say "transaction" is because sometimes when the app permission check finishes it doesn't always continue the app. Hence, it gets stranded, like the app is unresponsive but it's still functioning as the progress bar is still being displayed waiting till the permission check is finished... If that makes sense. That way I can detect when the app permission check finished and if the app got stuck on the end part of it or due to starting the next action in line.

So I have another question about transactions I am currently getting bookings every minute, these requests currently only have error handling. Is it a good idea to create a "Get bookings transaction" every minute or should I rather make a single transaction and add a metric to determine the number of successful requests that were made since the last error? The reason for this is that I want to be able to compare how many requests were successful and how many weren't. To predict the success rate of data requests for each method.

> What you describe looks like a use case for custom metrics for performance to me. You could attach both metrics you described as measurements to a transaction, see https://docs.sentry.io/platforms/android/performance/instrumentation/performance-metrics/.

That's a good idea, I didn't think of that (quite new to sentry so).

**Reply by philipphofmann** [reply]
> Well, the reason I say "transaction" is because sometimes when the app permission check finishes it doesn't always continue the app. Hence, it gets stranded, like the app is unresponsive but it's still functioning as the progress bar is still being displayed waiting till the permission check is finished... If that makes sense. That way I can detect when the app permission check finished and if the app got stuck on the end part of it or due to starting the next action in line.

If you can detect in your code that the permission check didn't succeed, I would rather send an event, with something like
```Java
Sentry.captureMessage("Permission check failed");
```

> Is it a good idea to create a "Get bookings transaction" every minute or should I rather make a single transaction and add a metric to determine the number of successful requests that were made since the last error?

Your transaction should represent a single unit of work. So, I would create one "Get bookings transaction" every time your code gets the bookings. Sentry automatically links errors happening during a transaction to that transaction, if the transaction is [bound to the scope](https://docs.sentry.io/platforms/java/performance/instrumentation/custom-instrumentation/#connect-errors-with-spans). I would just capture an error when the request fails, and let Sentry figure it out. If you are not satisfied with what you see in Sentry, you can still add custom metrics. I think reading this doc would help you  https://docs.sentry.io/product/sentry-basics/tracing/distributed-tracing/
